src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -242,10 +248,26 @@ public static void VisitScopeTree(Scope treeRoot, Action<Scope> action)
             /// visits the bound tree and translates information from the bound tree about
             /// variable scope, declared variables, and variable captures into the resulting
             /// <see cref="Scope"/> tree.
+            /// 
+            /// At the same time it sets <see cref="Scope.CanMergeWithParent"/>
+            /// for each Scope. This is done by looking for <see cref="BoundGotoStatement"/>s 
+            /// and <see cref="BoundConditionalGoto"/>s that jump from a point 
+            /// after the begining of a <see cref="Scope"/>, to a <see cref="BoundLabelStatement"/>
+            /// before the start of the scope, but after the start of <see cref="Scope.Parent"/>.
+            /// 
+            /// All control structures have been converted to gotos and labels by this stage,
```

Comment:
All _loops_ have been converted to . . .    
Technically there are still switches, try/throw in the tree. However, since you are looking for potential re-execution of lambda expressions without leaving their parent scopes, you only care about backward branches. And loops, which are conveniently lowered to backward branches when we reach here.